Add Sections 1.4 and 1.5 to A Level Computer Science (#7)
Draggie306/CheatSheets: A level/source/Computer Science.md

> Some numbers with decimals will never be accurately stored. An infinite level of precision will be required to represent numbers, using inverse powers of 2, such as 0.1 and 0.2. For example, 2/3 would be 0.1010101010...

### Floating point arithmetic
Floating point binary numbers allow numbers with decimal places to be represented. They are represented with a **mantissa** and an **exponent**, where the mantissa is the actual data of the number and the exponent is how much to shift this by.

Using the above example, 1100 1.111, we need to convert this into floating point arithmetic to ensure the computer can understand it as it doesn't know what a '.' is in this context.

We start by mentally moving the . to between the first two digits, so:

`1 1 0 0 1.1 1 1`
Becomes
`1.1 0 0 1 1 1 1`

However, this would now represent some decimal of 1. So now we need to specify an exponent. 

> In this example, as the decimal point moved 4 places, the exponent therefore becomes 4. We can represent this number as:

Reviewer: Rephrase this for clarification; "However, as this would now represent some decimal of 1, not the intended number, we need to specify an exponent."

`11001111 0100`

With an 8 bit mantissa and a 4 bit exponent.

#### Two's complement with Floating Point (Floating Point Normalisation)

> Because Two's complement relies on the most significant bit (first bit) being 1 for it to be negative, this has an unintended side-consequence with our number as the computer may think it is negative when we want it to be positive. 

Reviewer: Change (first bit) to (leftmost bit) and simplify phrasing: something like:

Because Two's complement relies on the most significant bit (leftmost bit) being `1` for negative numbers, we cannot represent positive numbers starting from `1`. To resolve this side-effect, we **normalise** the binary expression with the following rules:

In that case, an additional rule applies:
* for positive numbers, the imaginatory '.' needs to go between the first 0 and 1.
* for negative numbers, the imaginatory '.' needs to go between the first 1 and 0. 

So actually, in this number's case, as it is a positive number we need to add a trailing zero:

`0 1 1 0 0 1.1 1 1`

Now we can 'move' the . five places to the leading 0 and 1. 

`0.1 1 0 0 1 1 1 1`

This causes an exponent of 5, meaning the number is now:

`011001111 0101`

With an 9 bit mantissa and a 4 bit exponent.

#### Adding and Subtracting Floating Point

> We need to ensure to line up the exponents of the numbers to perform arithmetic operations on them. This may result in trailing zeros being added to both sides of the equation.

Reviewer: Looks good.

## 

  
# 1.5 Legal, moral, cultural and ethical issues

## Computing related legislation

> There are four laws that are specified. Typically, long-form questions are asked on these.

Reviewer: Maybe add that these are a hotbed for 9-12 mark questions?

* [Copyright, Designs & Patents Act 1988](https://www.legislation.gov.uk/ukpga/1988/48/contents)
* [Regulation of Investigatory Powers Act 2000](https://www.legislation.gov.uk/ukpga/2000/23/contents)

### Data Protection Act

The **Data Protection Act 2018** controls the way data about living people is stored and processed.
- It is a national law which complements the European Union's General Data Protection Regulation (GDPR).

There are six key principles of the Data Protection Act [Source](https://www.gov.uk/data-protection):
* data must be used fairly, transparently and lawfully;
* data must be used for explicit purposes set out;
* data used relevantly and only for what is necessary;
* data is to be kept accurate and up-to-date;
* data is kept no longer than necessary (typically 7 years);
* data must be kept securely.

In organisations, a designated individual must be appointed the **data controller**, who is responsible for handling the data of **data subjects**. Their responsibility includes reporting to the **Information Commissioner**, a 'non-departmental public body' responsible for enforcing the Data Protection Act. 

> A company may incorporate this by clearly setting their uses in a **privacy policy** on their website. For example, iBaguette may collect full name and date of birth used for identification and appropriate age content grouping, not collecting information such as National Insurance number, retained securely for a limited time of 2 years from last use. 

Reviewer: Haha I like this

2. Unauthorised access to data with intent to commit further offences.
3. Unauthorised attempts to impair (break or modify) the operation of a computer or data. 
  * 3A: Making, supplying or obtaining articles, such as programs or data, for use in the above offences.

These offences usually incur fines and imprisonment. There are multiple offences to ensure all bases are covered, for example there may be a case where someone gets unauthorised access to medical records, which may only grant a fine or a small period in prison. However, if said person were to distribute those private medical records and then supplied tooling to obtain them, this would land a more serious sentence as per the law. 

### Copyright, Designs and Patents Act

The **Copyright, Designs and Patents Act 1988** covers the copying or use of other people's work. This act means that copyright over work is **automatically granted** to the creator. This includes digital media e.g. software, but also physical media such as books. This also allows them to control how their work is used, mainly through the use of **software licences** granting permission to copy or modify the work. 

> There are some exclusions that are also defined, mainly around fair use for research or education, but the act mainly ensures that work is protected legally and action can be taken against unauthorised uses.

Reviewer: ```suggestion
There are some exclusions to this, mainly around **fair use for research and education**, but the Act mainly ensures that work is legally protected, so action can be taken against unauthorised uses.
```

> An example would be a developer creating some bespoke software, for example Caiusoft. The developer holds **automatic copyright** to the name and code of Caiusoft. However, they have made their software open source (see 1.1.2) and have issued a software licence allowing for the modification of their software as long as they are credited and the project remains open source.

Reviewer: Great!!

This act regulates the types of surveillance that can be carried out weighed based on both national security concern as well as people's rights to privacy. This surveillance can take the form of monitoring communications (such as email and phone) and informants. It also governs rules around encrypted data, as according to the law itself a request can only be made for the "decryption key" as long as it's proportionate and ncessary. 

In real life, this may be used by the police to try to access criminal's messages, as long as they have sufficient probable cause to have gain access - for example, if they have evidence of crimes being committed within those messages, instead of "randomly" suspecting criminal activity and attempting to intercept communication.

## Moral and ethical issues 

There are many opportunities and risks associated with many of the points below. We have highlighted a few examples of each that you could use but, in the exam as long as it's sensible and pertinent, you should be able to use any exam well. 
* Computers in the workforce:

>   * Opportunities: increasing productivity from typing, maths on spreadsheets.

Reviewer: ```suggestion
  * Opportunities: increasing productivity from typing, maths on spreadsheets removing calculation errors, easier collaboration on work tasks, greater flexibility.
```

  * Risks: training required, job losses in more manual sectors, up-front cost for equipment.
* Automated decision making:
  * Opportunities: no need to have a human check through tedious work, utilising algorithms instead which check for patterns.
  * Risks: these algorithms will need to be created and they may not be accurate, potentially missing critical issues (e.g. fraud). 
* Artificial intelligence:

>   * Opportunities: less reliance on humans for decision making, especially with Large Language Models allows increased productivity.

Reviewer: ```suggestion
  * Opportunities: less reliance on humans for decision making, large language models automating more medial/lower-level tasks increasing overall productivity
```

  * Risks: job losses from excessive automation, over-dependence on AI causing loss in critical thinking, easily misused.
* Environmental effects:
  * Opportunities: more energy efficient designs mean less overall consumption, recycling or upcycling.

>   * Risks: energy used to manufacture and run, electronic waste from disused machines. 

Reviewer: ```suggestion
  * Risks: energy used to manufacture and run, strain on the electricity grid with larger datacentres, electronic waste/pollution/toxic landfill from disused machines.
```

* Censorship and the Internet:
  * Opportunities: ensures that young people aren't being exposed to content not suitable for their age.
  * Risk: limiting freedom of expression and speech, people can't post whatever they want.
* Monitor behaviour:
  * Opportunities: safety and security of citizens, ensure criminals aren't accessing criminal websites.
  * Risks: privacy concerns with how data is collected and used, tools required to conduct surveillance. 
* Analyse personal information:
  * Opportunities: data from individuals can be used to improve services by tailoring their business utilising this data.
  * Risks: privacy, data misuse or potential security breaches where all data is leaked and can be accessed by hackers.
* Piracy and offensive communications:

>   * Opportunities: prevents theft of intellectual work from people so they don't lose out, keeps communications clean and prevents cyberbullying.

Reviewer: ```suggestion
  * Opportunities: prevents theft/unauthorised distribution of intellectual work from those who created it, keeps communications clean and prevents cyberbullying.
```

>   * Risks: easy threshold to be considered a "pirate" - can be accidental, platform or social "rules" may be ambiguous.

Reviewer: ```suggestion
  * Risks: easy threshold to be considered a "pirate" - can be accidental, platform or social "rules" may be ambiguous, debate concerning what is truly considered "free speech" and how this differs around the world.
```

* Layout, colour paradigms and character sets.

>   * Opportunities: ensures accessibility from a wide range of users (such as colour blind individuals or those who utilise screen readers).

Reviewer: ```suggestion
  * Opportunities: more accessible to a wider range of users (such as colour blind individuals or those who utilise screen readers).
```

>   * Risks: more development time required, some devices or browsers may not support different character sets (e.g. some devices may support ASCII up to 128 characters, others may support Unicode containing many more characters).

Reviewer: Great balance of pros and cons here, I might add a few more
